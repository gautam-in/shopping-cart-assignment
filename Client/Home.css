body,html{
    height : 100%;
    margin: 0;
}
header {
    display: flex;
    justify-content: space-between;
    width : 80%;
    margin-left : 10%;
    height : 100%;
}
.header-container{
   box-shadow: 0px 3px 10px rgb(0 0 0 / 0.2);
   height: 12%;
}
.head-nav-left{
    display: flex;
    justify-content: center;
    align-items: baseline;
}
.head-nav-left a> img{
    margin-right: 100px;
    height: 80px;
    margin-top: 12px;
}
.head-nav-left a{
    text-decoration: none;
    color : black;
    margin-left : 20px;
    cursor: pointer;
}
.head-nav-right {
    display: flex;
    flex-direction : column;
}
.head-nav-cart img{
    height : 30px;
    padding-right: 10px;
}
.head-nav-cart{
    margin-top: 20%;
    background-color: rgb(0 0 0 / 0.1);
    padding-left: 5px;
    width : 120px;
    height : 70px;
    display: flex;
    justify-content: center;
    align-items: center;
}
.head-nav-login{
    display: flex;
    justify-content: flex-end;
    font-size: 0.7rem;
}
.head-nav-login a{
    padding : 3px;
    cursor: pointer;
    text-decoration: none;
    color : black;
}

.carousel-container{
    position : relative;
    margin: 0 auto;
    border: none;
    width: 80%;
    display: flex;
}
  .carousel__item {
    height: 300px;
    padding: 1em;
    font-family: "Quicksand", sans-serif;
    font-weight: bold;
    font-size: 2em;
    color: #ffffff;
    display: none;
    box-shadow: 0px 1px 10px rgb(0 0 0 / 0.2);
  }
  .carousel__item img{
      width: 100%;
  }

  .carousel__item--selected {
    display: block;
  }

  .carousel__nav {
    width: 100%;
    padding: 20px 0;
    position: absolute;
    bottom: 0;
    left: 0;
    text-align: center;
  }

  .carousel__button {
    width: 10px;
    height: 10px;
    display: inline-block;
    background: rgba(124, 40, 40, 0.2);
    border-radius: 50%;
    margin: 0 5px;
    cursor: pointer !important;
    z-index : 100;
  }

  .carousel__button--selected {
    background: #d00156;
  }

main{
    display : flex;
    flex-direction : column;
}
.Home-container{
    display: flex;
    flex-direction: column;
    width : 80%;
    margin: 0 auto;
}
.category-item:nth-child(2n){
    flex-direction: row-reverse;
}

.category-item{
    display: flex;
    box-shadow: 0px 1px 10px rgb(0 0 0 / 0.2);
}
.cat-two{
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
}
.cat-two button{
    margin:30px 0;
    height: 30px;
    background-color: #ad3a6a;
    border : 0.2px solid rgb(46, 45, 45);
    color : white;
    cursor: pointer;
    width: max-content;
    padding : 0 10px;
}
.cat-two a{
    text-decoration: none;
    color : white;
}
footer{
    background-color:  rgb(0 0 0 / 0.1);
    height: 40px;
}
footer p{
    margin-left: 15%;
    padding-top: 10px;
}


@media screen and (max-width : 480px){
    .head-nav-left a:not(:first-child) {
        display: none;
    }
    .head-nav-left:first-child  img{
            height : 60px;
    }
    .head-nav-left:first-child {
        margin-left : 0;
    }
    .head-nav-left a{
        margin-left : 0;
    }
    .head-nav-login{
        display: none;
    }
    .head-nav-cart{
        margin-top: 20%;
        background-color: rgb(0 0 0 / 0.1);
        padding-left: 0;
        width: 100px;
        height: 50px;
        display: flex;
        justify-content: center;
        align-items: center;
    }
    .head-nav-cart img{
        height:  20px;
    }
    .carousel-container{
        height : 100px;
    }
    .carousel__item--selected {
        display: block;
    }
    .carousel__item {
        height: 100px;
        padding: 0;
    }
    .carousel__nav{
        top :  100px;
    }

    .cat-one img{
        width : 100px;
        height : 100px;
    }
    .category-item{
        align-items: center;
    }
    .cat-two button {
        margin: 20px 0;
        height: 20px;
        font-size: 0.5em;
    }
    .cat-two h3{
        font-size: 1em;
        margin-top: 50px;
    }
    .cat-two div{
        font-size: 0.5em;
    }
    footer p{
        font-size: 0.5em;
    }
    footer{
        justify-content: center;
    }
}

@media only screen
and (min-device-width : 768px)
and (max-device-width : 1024px)
and (orientation : landscape) {

    .head-nav-cart{
        margin-top: 10%;
        background-color: rgb(0 0 0 / 0.1);
        padding-left: 0;
        width: 100px;
        height: 50px;
        display: flex;
        justify-content: center;
        align-items: center;
    }
    .head-nav-cart img{
        height:  20px;
    }

}
