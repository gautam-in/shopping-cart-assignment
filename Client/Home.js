function FetchBanners() {
     return fetch("http://localhost:5000/banners").then((res) => {
      if (res.ok) {
            return res.json();

      } else {
        throw new Error("Something went wrong !");
      }
    })
  };

  function FetchCategories() {
     return fetch("http://localhost:5000/categories").then((res) => {
      if (res.ok) {
            return res.json();

      } else {
        throw new Error("Something went wrong !");
      }
    })
  };

  async function DisplayBanners() {
    let allBanners = await FetchBanners();
    for (const banner of allBanners) {
      BannercardItem(banner);
    }
    addBannerButtons(allBanners.length);
    addButtonFunctionality();
    DisplayCategories();
  }


  async function DisplayCategories() {
    let allCategories = await FetchCategories();
    for (const cat of allCategories) {
      CategoryCardItem(cat);
    }
  }

window.addEventListener("DOMContentLoaded", DisplayBanners);

function BannercardItem(banner){
    let bannerItem = document.createElement("div");
    bannerItem.setAttribute("class","carousel__item");
    bannerItem.setAttribute("id",`..${banner.id}`);
    bannerItem.setAttribute("order",`..${banner.order}`);
    bannerItem.setAttribute("isActive",`..${banner.isActive}`);

    let bannerImage=document.createElement("img");
    bannerImage.setAttribute("alt",`..${banner.bannerImageAlt}`);
    bannerImage.setAttribute("src",`..${banner.bannerImageUrl}`);

    bannerItem.appendChild(bannerImage);

    document.querySelector(".carousel-container").appendChild(bannerItem);

}

function addBannerButtons(len){
    let carouselButtonContainer = document.createElement("div");
    carouselButtonContainer.setAttribute("class","carousel__nav");
    document.querySelector(".carousel-container").appendChild(carouselButtonContainer)

    for(i=0;i<len;i++){
        let bannerNavButton= document.createElement("span");
        bannerNavButton.setAttribute("class","carousel__button");
        document.querySelector(".carousel__nav").appendChild(bannerNavButton);
    }


}

function addButtonFunctionality(){
    document.querySelectorAll(".carousel-container").forEach((carousel) => {
        const items = document.querySelectorAll(".carousel__item");
        const buttons = carousel.querySelectorAll(".carousel__button");

        buttons.forEach((button, i) => {
          button.addEventListener("click", () => {

            items.forEach((item) =>{

                item.classList.remove("carousel__item--selected")
            }

            );
            buttons.forEach((button) =>
              button.classList.remove("carousel__button--selected")
            );

            items[i].classList.add("carousel__item--selected");
            button.classList.add("carousel__button--selected");
          });
        });
        console.log()
        items[0].classList.add("carousel__item--selected");
        buttons[0].classList.add("carousel__button--selected");
      });

}

function CategoryCardItem(cat){
    if(cat.enabled){
          let catItem = document.createElement("div");
          catItem.setAttribute("class","category-item");

          let catOne=document.createElement("div");
          catOne.setAttribute("class","cat-one");

          let catTwo=document.createElement("div");
          catTwo.setAttribute("class","cat-two");

          let catImage=document.createElement("img");
          catImage.setAttribute("alt",`${cat.key}`);
          catImage.setAttribute("src",`..${cat.imageUrl}`);


          let catButton = document.createElement("Button");
          let catLink = document.createElement("a");
          catLink.setAttribute("href","./Products.html");
          catButton.appendChild(catLink);
          catLink.innerHTML= `Explore ${cat.key}`;
          catButton.setAttribute("href","Products.html");

          let catDescription = document.createElement("div");
          catDescription.innerHTML= `${cat.description}`;

          let catHeader = document.createElement("h3");
          catHeader.textContent= `${cat.name}`;

          catItem.appendChild(catOne);
          catItem.appendChild(catTwo);
          catOne.appendChild(catImage);
          catTwo.appendChild(catHeader);
          catTwo.appendChild(catDescription);
          catTwo.appendChild(catButton);

          document.querySelector(".Home-container").appendChild(catItem);
    }

}
