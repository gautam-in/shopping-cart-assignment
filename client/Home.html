<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Sabka Bazar</title>
    <link rel="stylesheet" href="style/design.css" />
  </head>

  <body>
    <header id="main-header" class="main-header">
      <uc-header count="0"></uc-header>
    </header>
    <div class="banners-container" id="banners-container"></div>
    <footer>
      <uc-footer></uc-footer>
    </footer>
  </body>
  <script defer src="./header.js"></script>
  <script defer src="./footer.js"></script>
  <script defer>
    const baseUrl = "../../";
    function getProductsList() {
      const bannerPromise = fetch("http://localhost:5000/banners").then((res) =>
        res.ok ? res.json() : ""
      );
      const categoriesPromise = fetch("http://localhost:5000/categories").then(
        (res) => (res.ok ? res.json() : "")
      );
      Promise.all([bannerPromise, categoriesPromise])
        .then((response) => {
          showCategoriesList(response[1]);
        })
        .catch((error) => {
          console.log("Unexpected Error");
        });
    }

    function showCategoriesList(categories) {
      const container = document.querySelector("#banners-container");
      categories.forEach((category, index) => {
        container.appendChild(
          showCategoryData(
            category,
            index % 2 === 0 ? "imageRight" : "imageLeft"
          )
        );
      });
    }

    function showCategoryData(category, imageAlignment) {
      let categoryCard = document.createElement("div");
      categoryCard.classList.add("category-banner");

      let GoodsTypeInfo = document.createElement("div");
      GoodsTypeInfo.classList.add("category-details");

      let categoryImage;
      if (category.imageUrl) {
        GoodsImgs = document.createElement("div");
        GoodsImgs.classList.add("image-section");
        categoryImage = document.createElement("img");
        categoryImage.setAttribute("src", `${baseUrl}${category.imageUrl}`);
        categoryImage.setAttribute("alt", category.name);
        GoodsImgs.append(categoryImage);
      } else {
        categoryImage = document.createElement("div");
      }
      let categoryName = document.createElement("h3");
      categoryName.classList.add("category-name");
      categoryName.innerHTML = category.name;

      let categoryDescription = document.createElement("div");
      categoryDescription.classList.add("category-des");
      categoryDescription.innerHTML = category.description;

      let categoryBtn = document.createElement("button");
      categoryBtn.classList.add("category-btn");
      categoryBtn.innerHTML = `Explore ${category.key}`;
      categoryBtn.addEventListener("click", () =>
        goToCategorySection(category.id)
      );

      GoodsTypeInfo.append(categoryName);
      GoodsTypeInfo.append(categoryDescription);
      GoodsTypeInfo.append(categoryBtn);

      if (imageAlignment === "imageLeft") {
        categoryCard.append(GoodsImgs);
        categoryCard.append(GoodsTypeInfo);
      } else {
        categoryCard.append(GoodsTypeInfo);
        categoryCard.append(GoodsImgs);
      }

      return categoryCard;
    }

    function goToCategorySection(categoryId) {
      location.href = `./product.html?id=${categoryId}`;
    }

    getProductsList();
  </script>
</html>
