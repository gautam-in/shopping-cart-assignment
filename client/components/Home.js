import { useEffect,useState } from "react";
import { CategoryArticle } from "../styles/HomeStyle";
import "react-responsive-carousel/lib/styles/carousel.min.css";
import { Carousel } from "react-responsive-carousel";
import HomeCategory from "./HomeCategory";
import { getData } from "../utility/Datahandler";
import Error from "./Error";

const Home = () => {
  const [bannerList, setBannerList] = useState([]);

  useEffect(() => {
    if (bannerList.length === 0) getData("/banners")
      .then((response) => {
        setBannerList(response);
      })
  }, []);

  return (
    <>
      {bannerList.length === 0 ? (
        <Error errorMessage="Unable to load banners." />
      ) : (
        <CategoryArticle>
          <Carousel
            preventMovementUntilSwipeScrollTolerance
            swipeable
            showStatus={false}
            showThumbs={false}
            emulateTouch
            showArrows={true}
          >
            {bannerList
              ?.filter((banner) => banner.isActive)
              .map((banner) => (
                <div key={banner.id}>
                  <img
                    src={banner.bannerImageUrl}
                    alt={banner.bannerImageAlt}
                  />
                </div>
              ))}
          </Carousel>
        </CategoryArticle>
      )}
      <HomeCategory />
    </>
  );
};

export default Home;
