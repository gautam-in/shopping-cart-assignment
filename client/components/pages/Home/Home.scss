@import "/client/styles/mq-mixin";

.home_container {
  margin-top: 20px;
  > div:first-child {
    box-shadow: 0px 10px 11px -12px #c1c1c1;
    margin-bottom: 20px;
    padding: 10px 0px;
  }

  .explore {
    margin: 10px;
  }
}

.gradient_shadow {
  height: 20px;
  background-image: linear-gradient(
      to right,
      #ffffff,
      #ffffff,
      #ffffff,
      #bcbcbc,
      #bcbcbc,
      #bcbcbc,
      #bcbcbc,
      #ffffff,
      #ffffff,
      #ffffff
    ),
    linear-gradient(to bottom, #b7b7b7, #c9c9c9, #ffffff, #ffffff, #ffffff);
}
