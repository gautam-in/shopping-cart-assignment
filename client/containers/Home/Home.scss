@media only screen and (min-width: 1030px) {
    .home {
        padding: 0 6rem;
    }
}

$primary-color:#d00156;
$accent-color:#ea1d71;

.home {
    min-height: 80vh;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
}

.category-list {
    margin: 0;
    padding: 0;
    list-style-type: none;

    .category {
        display: flex;
        padding: 20px 10px 30px 10px;
        height: 10em;
        position: relative;
    }

    .category:after {
        content: '';
        position: absolute;
        bottom: 0;
        width: 90%;
        left: 5%;
        height: 10px;
        z-index: -1;
        background: radial-gradient(at 50% 0, rgba(0, 0, 0, 0.5), transparent 70%);
    }

    .category:nth-child(even) {
        flex-direction: row-reverse;
    }

    .category-details {
        display: flex;
        justify-content: space-evenly;
        align-items: center;
        flex-direction: column;
        width: 100%;

        .category-title {
            font-weight: bold;
            font-size: 20px;
            text-align: center;
        }

        .category-description {
            text-align: center;
        }

        .category-explore-button {
            padding: 10px;
            background-color: $primary-color;
            border-radius: 0px;
            color: white;
            border: none;
            cursor: pointer;

            &:hover,
            &:focus {
                background: $accent-color;
            }
        }

        button:disabled {
            background: gray;
        }

        .category .category-explore-button a {
            color: white;
            text-decoration: none;
        }
    }

    .category-right {
        display: flex;
        align-items: center;
        width: 35%;

        .category-image {
            width: 100%;
            height: auto;
        }
    }
}