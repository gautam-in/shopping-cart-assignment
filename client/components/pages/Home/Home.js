import React, { useEffect, useState } from "react";
import { useHistory } from "react-router";

import Carousel from "../../organisms/Carousel/Carousel";
import Explore from "../../organisms/Explore/Explore";

import * as service from "../services";

import "./Home.scss";

function Home() {
  const history = useHistory();
  const [offers, setOffers] = useState([]);
  const [categories, setCategories] = useState([]);

  const getOffers = () => {
    service
      .getOffersForCarousel()
      .then((data) => {
        setOffers(data);
      })
      .catch((err) => {
        console.log("ERROR detected fetching products", err);
      });
  };
  const getCategoriesOptions = () => {
    service
      .getCategories()
      .then((data) => {
        let temp = data.filter((ele) => ele.enabled);
        temp.sort((a, b) => a.order - b.order);
        setCategories(temp);
      })
      .catch((err) => {
        console.log("ERROR detected fetching products", err);
      });
  };

  useEffect(() => {
    getOffers();
    getCategoriesOptions();
  }, []);

  return (
    <div className="home_container">
      <div>
        <Carousel offers={offers} />
      </div>
      {/* To-do Gradient bottom */}
      {/* <div className="gradient_shadow"></div> */}
      <div className="explore">
        {categories.map((category, i) => (
          <Explore key={category.id} index={i} category={category} />
        ))}
      </div>
    </div>
  );
}

export default Home;
