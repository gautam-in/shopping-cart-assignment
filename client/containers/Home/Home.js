import React, { useEffect } from "react";
import "./Home.scss";
import { useHistory } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";
import { fetchCategories } from "../../redux/categories/categoryActions";
import Carousel from "../../components/Carousel/Carousel";
import { fetchBanners } from "../../redux/banners/bannerActions";

const Home = () => {
  const history = useHistory();
  const dispatch = useDispatch();
  const categories = useSelector((state) => state.categories.data);
  const loading = useSelector((state) => state.categories.loading);
  const error = useSelector((state) => state.categories.error);
  const banners = useSelector((state) => state.banners.data);

  useEffect(() => {
    dispatch(fetchBanners());
    dispatch(fetchCategories());
  }, []);

  const handleExplore = (categoryId) => {
    history.push(`/products#${categoryId}`);
  };

  return (
    <div className="home">
      {loading ? (
        <div className="no-content">Loading...</div>
      ) : error ? (
        <div className="no-content">Some error occured!</div>
      ) : (
        <>
          {banners && <Carousel data={banners} />}
          <ul className="category-list">
            {categories.map((category) => (
              <li key={category.id} className="category">
                <div className="category-details">
                  <div className="category-title">{category.name}</div>
                  <div className="category-description">
                    {category.description}
                  </div>
                  <button
                    type="button"
                    className="category-explore-button"
                    onClick={() => handleExplore(category.id)}
                    tabIndex={0}
                    disabled={!category.enabled}
                    onKeyPress={() => handleExplore(category.id)}
                  >
                    Explore {category.name}
                  </button>
                </div>
                <div className="category-right">
                  <img
                    className="category-image"
                    src={category.imageUrl}
                    alt={category.name}
                    height="150"
                    width="200"
                  />
                </div>
              </li>
            ))}
          </ul>
        </>
      )}
    </div>
  );
};

export default Home;
