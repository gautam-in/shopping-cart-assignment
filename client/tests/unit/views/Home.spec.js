import { mount, shallowMount } from "@vue/test-utils";
import Home from "@/views/Home.vue";
import Categories from "@/components/Categories.vue";
import Vue from "vue";
import VueRouter from "vue-router";
const elem = document.createElement("div");
if (document.body) {
  document.body.appendChild(elem);
}
describe("Home", () => {
  it("snapshot", () => {
    const wrapper = mount(Home, {
      propsData: {
        banners: [],
        categories: [],
      },
      stubs: {
        Categories,
      },
      attachTo: elem,
    });
    expect(wrapper).toMatchSnapshot();
    wrapper.destroy();
    // expect(wrapper.contains("div")).toBe(true);
  });
});
