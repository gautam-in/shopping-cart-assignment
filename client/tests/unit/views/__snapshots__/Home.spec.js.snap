// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home snapshot 1`] = `
<div class="home" banners="" categories="">
  <!---->
</div>
`;
